rustls/src/server/tls12.rs

```diff
@@ -998,7 +999,30 @@ impl State<ServerConnectionData> for ExpectTraffic {
             .extract_secrets(Side::Server)
     }
 
+    fn into_external_state(self: Box<Self>) -> Result<Box<dyn ExternalState + 'static>, Error> {
+        Ok(self)
+    }
+
     fn into_owned(self: Box<Self>) -> hs::NextState<'static> {
         self
     }
 }
+
+impl ExternalState for ExpectTraffic {
+    fn update_secrets(&mut self, _: Direction) -> Result<ConnectionTrafficSecrets, Error> {
+        Err(Error::General(
+            "TLS 1.2 connections do not support traffic secret updates".into(),
+        ))
+    }
+
+    fn handle_new_session_ticket(
+        &mut self,
+        _cx: &mut ExternalContext<'_>,
+        _message: &NewSessionTicketPayloadTls13,
+    ) -> Result<(), Error> {
+        Err(Error::InappropriateMessage {
```

Comment:
Don't we have a better error variant for wrong protocol version? If not, maybe we should create one?

Reply:
This gets handled higher up in https://github.com/swlynch99/rustls/blob/external-connection/rustls/src/conn/kernel.rs#L226. I'm happy to change the error here, but this is dead code so I just mirrored what rustls returns in this case for a normal connection.

Reply:
I don't mind that; maybe we could put a `&'static str` in there so we can work out which one was incorrect, if someone who reports seeing such an error.

Reply:
Here's what I have done instead:
- For `server::tls12::ExpectTraffic` I have added a useful error message that matches what `KernelConnection::handle_new_session_ticket` returns
- For `client::tls1[23]::ExpectTraffic` I have put `unreachable!` with a useful error message. I have also added a note that the method is only really supposed to be called for client connections.

All three are dead code, but the client ones should definitely never be called (and I've added a note to the `KernelState` documentation about this) whereas the server-side of the connection should just be straight-up unreachable.